lightning/src/routing/scoring.rs

```diff
@@ -997,10 +997,30 @@ impl<L: Deref<Target = u64>, BRT: Deref<Target = HistoricalBucketRangeTracker>,
 	}
 
 	fn decayed_offset_msat(&self, offset_msat: u64) -> u64 {
-		self.now.duration_since(*self.last_updated).as_secs()
-			.checked_div(self.params.liquidity_offset_half_life.as_secs())
-			.and_then(|decays| offset_msat.checked_shr(decays as u32))
-			.unwrap_or(0)
+		let elapsed_time = self.now.duration_since(*self.last_updated).as_secs();
+		let half_life = self.params.liquidity_offset_half_life.as_secs();
+
+		match elapsed_time.checked_div(half_life) {
+			None => 0,
+			Some(decays) => match elapsed_time.checked_div(half_life / 2) {
+				None => 0,
+				Some(half_decays) => {
+					// Decay the offset by the appropriate number of half lives. If half of the next
```

Comment:
why are we doing this in such a complicated manner if we're trying to increase precision? We can do things much simpler if we permit floating points:

double halving_count = elapsed_time/half_life
double factor = 0.5^halving_count
decayed_msat = int(factor * offset_msat)

Reply:
This isn't about increasing precision -- right shift divides by two so there would only be half msat loss of precision. It's about having less noticeable jumps between two half lives.

Reply:
Ah, I see. Use floating point division. I can benchmark though previously we wanted to avoid floating-point operations.

Reply:
I think you should give it a try. It's the easiest way to avoid these discrete jumps and smooth out the decay.

Reply:
> Also the discussion on float-vs-intop doesn't make sense to me? Using floats here isn't increasing precision, the change is the absolute calculation using `exp` and `div`, which are both potentially very slow. The above calculation could happen just as well with fixed-point and with almost identical precision.

I think the idea is `halving_count` would be a floating point number whereas with integer division it results in having discrete, interger-by-integer steps.

Reply:
> I think you should give it a try. It's the easiest way to avoid these discrete jumps and smooth out the decay.

It's about 0.5ms slower by my benchmark. Added a 10s sleep before running to ensure a non-zero elapsed time.

> Division is _pain_, I don't think we should divide here, if at all possible. Sadly, just benchmarking it on x86 isn't really sufficient, IMO, we'll want to benchmark, at least, on (a) mobile, ideally both Apple and samsung/more-defualt ARM, (b) WASM, again ideally both Apple/ARM and x86, (c) few-generations-old x86. If we really want to go this route, we should consider just using a timer, see discussion at #1752, which would remove all the complexity or need to consider the performance here.

FWIW, making the decay an identity function shaved 3-4ms off the benchmark on my machine. MacBook Pro 2.4 GHz Quad-Core Intel Core i5.

Reply:
What are your thoughts on the original approximation approach? That didn't have any difference.

Reply:
Need is a strong word. 😛  There is a tradeoff with the added dependency. I'm not entirely convinced it's warranted yet at least without evidence of a problem.

Reply:
Managed to get `cargo flamegraph` working on it. Looks like one of the biggest wins would be to avoid hitting the `Instant::now` and `Instant::duration_since` for each channel. Seeing time spent in `libsystem_kernel.dylib`mach_absolute_time` a few places in the `channel_penalty_msat` call tree.  Attached are couple graphs. Had to force some code to never inline.

<img width="2552" alt="get_route" src="https://user-images.githubusercontent.com/4015903/197623702-14a6d3a9-06ee-4349-a9cc-5d56a6dcf593.png">
<img width="2546" alt="channel_penalty_msat" src="https://user-images.githubusercontent.com/4015903/197623710-f47aa001-b36b-4d38-a348-d24144540fc3.png">

Could perhaps wrap the `ProbabilisticScorer` in such a way that we're only getting the time once.

Reply:
Here's the actual SVG file.

![flamegraph](https://user-images.githubusercontent.com/4015903/197626108-d0ef6b92-401e-4a39-94c6-1c623e5c66a5.svg)

Reply:
Whelp... github managed to make it non-interactive. 😢

Reply:
Swapped in `SinceEpoch` and advanced the time 30 hours (4.5 decays) between initial setup and running the benchmark and it took 38-39ms over three runs. I think the simplest approach is approximating to smooth out the decay and fixing `Time::now` before `find_route` to improve performance, if needed. AFAICT, the clock is main culprit, not the operations needed to compute the decay.

Reply:
Without my change, 35-38ms across a few runs compared to 17-19ms with zero penalty scorer. Both about 10ms faster than baseline. Simulating a fixed time with `SinceEpoch` brings it down to 25-28ms.

Reply:
Getting back to this, seeing 30-32ms when running with your branch and using an identity function for `decayed_offset_msat`. Same take away as before.

Reply:
Even with the identity function, there's still a call to `T::now()` in `as_directed`. When `T` is `SinceEpoch`, `T::now()` is just wrapping a `Duration`.